File: frontend/src/lib/api/download.ts
/** Download API types and functions */
import { apiClient } from './client';

export interface DataProductInfo {
	access_url: string;
	uid: string;
	filename: string;
	content_length: number;
	content_type: string;
	product_type: string;
	size_mb: number;
}

export interface ResolveProductsResponse {
	products: DataProductInfo[];
	total_count: number;
	total_size_bytes: number;
	total_size_display: string;
	by_type: Record<string, { count: number; size_bytes: number; size_display: string }>;
}

export interface DiskSpaceInfo {
	path: string;
	total_bytes: number;
	used_bytes: number;
	free_bytes: number;
	total_display: string;
	used_display: string;
	free_display: string;
	sufficient: boolean;
}

export interface StartDownloadResponse {
	job_id: string;
	status: string;
	total_files: number;
	total_bytes: number;
	total_size_display: string;
	destination: string;
}

export interface FileStatus {
	filename: string;
	content_length: number;
	bytes_downloaded: number;
	status: string;
	error: string | null;
	progress: number;
}

export interface DownloadJobStatus {
	job_id: string;
	status: string;
	destination: string;
	total_files: number;
	files_completed: number;
	files_failed: number;
	total_bytes: number;
	bytes_downloaded: number;
	progress: number;
	error: string | null;
	files: FileStatus[];
}

export interface DownloadJobSummary {
	job_id: string;
	status: string;
	destination: string;
	total_files: number;
	files_completed: number;
	files_failed: number;
	progress: number;
	created_at: string;
	member_ous_uids: string[];
	product_filter: string;
	total_bytes: number;
	bytes_downloaded: number;
	error: string | null;
}

export interface BrowseDirectoryEntry {
	name: string;
	path: string;
	is_dir: boolean;
}

export interface BrowseDirectoryResponse {
	current: string;
	parent: string | null;
	entries: BrowseDirectoryEntry[];
}

export const downloadApi = {
	/** Browse server-side directories */
	browseDirectory(path: string = '/'): Promise<BrowseDirectoryResponse> {
		return apiClient.get<BrowseDirectoryResponse>(
			`/api/v1/downloads/browse?path=${encodeURIComponent(path)}`
		);
	},

	/** Create a directory and return its browse result */
	createDirectory(path: string): Promise<BrowseDirectoryResponse> {
		return apiClient.post<BrowseDirectoryResponse>(
			`/api/v1/downloads/mkdir?path=${encodeURIComponent(path)}`
		);
	},

	/** Resolve downloadable products for selected member_ous_uids */
	resolveProducts(memberOusUids: string[]): Promise<ResolveProductsResponse> {
		return apiClient.post<ResolveProductsResponse>('/api/v1/downloads/resolve', {
			member_ous_uids: memberOusUids
		});
	},

	/** Check disk space at a given path */
	checkDiskSpace(path: string, neededBytes: number = 0): Promise<DiskSpaceInfo> {
		return apiClient.post<DiskSpaceInfo>('/api/v1/downloads/disk-space', {
			path,
			needed_bytes: neededBytes
		});
	},

	/** Start a download job */
	startDownload(params: {
		memberOusUids: string[];
		productFilter?: string;
		destination: string;
		maxParallel?: number;
		extractTar?: boolean;
	}): Promise<StartDownloadResponse> {
		return apiClient.post<StartDownloadResponse>('/api/v1/downloads/start', {
			member_ous_uids: params.memberOusUids,
			product_filter: params.productFilter ?? 'all',
			destination: params.destination,
			max_parallel: params.maxParallel ?? 3,
			extract_tar: params.extractTar ?? false
		});
	},

	/** List all download jobs */
	listJobs(): Promise<DownloadJobSummary[]> {
		return apiClient.get<DownloadJobSummary[]>('/api/v1/downloads/jobs');
	},

	/** Get detailed status for a specific job */
	getJobStatus(jobId: string): Promise<DownloadJobStatus> {
		return apiClient.get<DownloadJobStatus>(`/api/v1/downloads/jobs/${encodeURIComponent(jobId)}`);
	},

	/** Cancel a download job */
	cancelJob(jobId: string): Promise<{ message: string }> {
		return apiClient.post<{ message: string }>(
			`/api/v1/downloads/jobs/${encodeURIComponent(jobId)}/cancel`
		);
	},

	/** Re-download a previous job using stored file records */
	redownloadJob(jobId: string): Promise<StartDownloadResponse> {
		return apiClient.post<StartDownloadResponse>(
			`/api/v1/downloads/jobs/${encodeURIComponent(jobId)}/redownload`
		);
	},

	/** Delete a download job from history */
	deleteJob(jobId: string): Promise<{ job_id: string; deleted: boolean }> {
		return apiClient.delete<{ job_id: string; deleted: boolean }>(
			`/api/v1/downloads/jobs/${encodeURIComponent(jobId)}`
		);
	}
};

File: backend/app/schemas/download.py
"""Download-related schemas."""

from typing import Any, List, Optional

from pydantic import BaseModel, Field


class DataProductInfo(BaseModel):
    """A single downloadable data product."""

    access_url: str
    uid: str = Field(..., description="member_ous_uid")
    filename: str
    content_length: int = Field(..., description="Size in bytes")
    content_type: str = ""
    product_type: str = Field(
        ...,
        description=(
            "Classified type: all, raw, calibration, scripts, weblog, qa_reports, "
            "auxiliary, cubes, continuum, fits, other"
        ),
    )
    size_mb: float = Field(..., description="Size in MB")


class BrowseDirectoryEntry(BaseModel):
    """A single entry in a directory listing."""

    name: str
    path: str
    is_dir: bool


class BrowseDirectoryResponse(BaseModel):
    """Response for browsing server-side directories."""

    current: str = Field(..., description="Resolved absolute path of the directory")
    parent: Optional[str] = Field(None, description="Parent directory path, None if at root")
    entries: List[BrowseDirectoryEntry]


class ResolveProductsRequest(BaseModel):
    """Request to list downloadable products for selected observations."""

    member_ous_uids: List[str] = Field(
        ..., description="List of member_ous_uid values to resolve products for"
    )


class ResolveProductsResponse(BaseModel):
    """Response listing available products and size breakdown."""

    products: List[DataProductInfo]
    total_count: int
    total_size_bytes: int
    total_size_display: str = Field(..., description="Human-friendly size string")
    by_type: dict = Field(
        default_factory=dict,
        description="Breakdown: {type: {count, size_bytes, size_display}}",
    )


class DiskSpaceInfo(BaseModel):
    """Disk space info for a given path."""

    path: str
    total_bytes: int
    used_bytes: int
    free_bytes: int
    total_display: str
    used_display: str
    free_display: str
    sufficient: bool = Field(..., description="Whether free space >= needed bytes")


class CheckDiskSpaceRequest(BaseModel):
    """Request to check disk space for a destination."""

    path: str = Field(..., description="Destination directory path")
    needed_bytes: int = Field(0, description="Bytes that will be downloaded")


class StartDownloadRequest(BaseModel):
    """Request to start downloading selected products."""

    member_ous_uids: List[str] = Field(..., description="List of member_ous_uid values")
    product_filter: str = Field(
        "all",
        description=(
            "Product type filter: all, raw, calibration, scripts, weblog, qa_reports, "
            "auxiliary, cubes, continuum, fits, other"
        ),
    )
    destination: str = Field(..., description="Local directory to save files")
    max_parallel: int = Field(
        3,
        description="Maximum parallel download threads",
        ge=1,
        le=8,
    )
    extract_tar: bool = Field(
        False,
        description="Extract tar archives after download and remove the archives",
    )
    unpack_ms: bool = Field(
        False,
        description=("Create raw MeasurementSets from downloaded/extracted ALMA ASDM products"),
    )
    generate_calibrated_visibilities: bool = Field(
        False,
        description=(
            "Create calibrated split MeasurementSets from raw MS and delivered calibration scripts"
        ),
    )
    clean_intermediate_files: bool = Field(
        False,
        description=(
            "After calibrated MS creation, delete downloaded originals "
            "and intermediate raw MS products"
        ),
    )
    archive_output_root: Optional[str] = Field(
        None,
        description="Optional root directory for raw_ms and calibrated_ms outputs",
    )
    casa_data_root: Optional[str] = Field(
        None,
        description="Optional CASA runtime data directory",
    )
    skip_casa_data_update: bool = Field(
        False,
        description="Skip CASA runtime data updates",
    )
    selected_metadata: Optional[List[dict[str, Any]]] = Field(
        None,
        description="Metadata rows associated with this download job",
    )


class RedownloadRequest(BaseModel):
    """Request to re-download a previous job with new execution options."""

    max_parallel: Optional[int] = Field(
        None,
        description="Maximum parallel download threads",
        ge=1,
        le=8,
    )
    extract_tar: Optional[bool] = Field(
        None,
        description="Extract tar archives after download and remove the archives",
    )
    unpack_ms: Optional[bool] = Field(
        None,
        description="Create raw MeasurementSets after re-download",
    )
    generate_calibrated_visibilities: Optional[bool] = Field(
        None,
        description="Create calibrated split MeasurementSets after re-download",
    )
    clean_intermediate_files: Optional[bool] = Field(
        None,
        description="After calibrated MS creation, keep only split calibrated products",
    )
    archive_output_root: Optional[str] = Field(
        None,
        description="Optional root directory for raw_ms and calibrated_ms outputs",
    )
    casa_data_root: Optional[str] = Field(
        None,
        description="Optional CASA runtime data directory",
    )
    skip_casa_data_update: Optional[bool] = Field(
        None,
        description="Skip CASA runtime data updates",
    )


class StartDownloadResponse(BaseModel):
    """Response after initiating a download job."""

    job_id: str
    status: str
    total_files: int
    total_bytes: int
    total_size_display: str
    destination: str


class FileStatus(BaseModel):
    """Status of a single file in a download job."""

    filename: str
    content_length: int
    bytes_downloaded: int
    status: str
    error: Optional[str] = None
    progress: float = Field(..., description="Download progress 0.0 to 1.0")


class DownloadJobStatus(BaseModel):
    """Status of an entire download job."""

    job_id: str
    status: str
    destination: str
    total_files: int
    files_completed: int
    files_failed: int
    total_bytes: int
    bytes_downloaded: int
    progress: float = Field(..., description="Overall progress 0.0 to 1.0")
    error: Optional[str] = None
    files: List[FileStatus] = []
    raw_measurement_sets: List[str] = []
    calibrated_measurement_sets: List[str] = []
    manifest_path: Optional[str] = None
    has_metadata: bool = False
    metadata_count: int = 0
    unpack_ms: bool = False
    generate_calibrated_visibilities: bool = False
    clean_intermediate_files: bool = False
    archive_output_root: Optional[str] = None
    casa_data_root: Optional[str] = None
    skip_casa_data_update: bool = False


class DownloadJobSummary(BaseModel):
    """Summary of a download job for listing."""

    job_id: str
    status: str
    destination: str
    total_files: int
    files_completed: int
    files_failed: int
    progress: float
    created_at: str
    member_ous_uids: List[str] = []
    product_filter: str = "all"
    total_bytes: int = 0
    bytes_downloaded: int = 0
    error: Optional[str] = None
    raw_measurement_sets: List[str] = []
    calibrated_measurement_sets: List[str] = []
    manifest_path: Optional[str] = None
    has_metadata: bool = False
    metadata_count: int = 0
    unpack_ms: bool = False
    generate_calibrated_visibilities: bool = False
    clean_intermediate_files: bool = False
    archive_output_root: Optional[str] = None
    casa_data_root: Optional[str] = None
    skip_casa_data_update: bool = False
